contracts/core/EntryPoint.sol

```diff
@@ -753,75 +781,60 @@ contract EntryPoint is IEntryPoint, StakeManager, NonceManager, ReentrancyGuardT
      * @param mode      - Whether is called from innerHandleOp, or outside (postOpReverted).
      * @param opInfo    - UserOp fields and info collected during validation.
      * @param context   - The context returned in validatePaymasterUserOp.
-     * @param actualGas - The gas used so far by this user operation.
-     *
-     * @return actualGasCost - the actual cost in eth this UserOperation paid for gas
+     * @param executionGas - The gas used for execution of this user operation.
      */
     function _postExecution(
         IPaymaster.PostOpMode mode,
         UserOpInfo memory opInfo,
         bytes memory context,
-        uint256 actualGas
-    ) internal virtual returns (uint256 actualGasCost) {
+        uint256 executionGas
+    ) internal virtual returns (uint256 actualGas) {
         uint256 preGas = gasleft();
         unchecked {
-            address refundAddress;
             MemoryUserOp memory mUserOp = opInfo.mUserOp;
             uint256 gasPrice = getUserOpGasPrice(mUserOp);
 
             address paymaster = mUserOp.paymaster;
             // Calculating a penalty for unused execution gas
-            {
-                uint256 executionGasUsed = actualGas - opInfo.preOpGas;
-                // this check is required for the gas used within EntryPoint and not covered by explicit gas limits
-                actualGas += _getUnusedGasPenalty(executionGasUsed, mUserOp.callGasLimit);
-            }
-            uint256 postOpUnusedGasPenalty;
-            if (paymaster == address(0)) {
-                refundAddress = mUserOp.sender;
-            } else {
-                refundAddress = paymaster;
-                if (context.length > 0) {
-                    actualGasCost = actualGas * gasPrice;
-                    uint256 postOpPreGas = gasleft();
-                    if (mode != IPaymaster.PostOpMode.postOpReverted) {
-                        try IPaymaster(paymaster).postOp{
-                            gas: mUserOp.paymasterPostOpGasLimit
-                        }(mode, context, actualGasCost, gasPrice)
-                        // solhint-disable-next-line no-empty-blocks
-                        {} catch {
-                            bytes memory reason = Exec.getReturnData(REVERT_REASON_MAX_LEN);
-                            revert PostOpReverted(reason);
-                        }
-                    }
-                    // Calculating a penalty for unused postOp gas
-                    uint256 postOpGasUsed = postOpPreGas - gasleft();
-                    postOpUnusedGasPenalty = _getUnusedGasPenalty(postOpGasUsed, mUserOp.paymasterPostOpGasLimit);
+            actualGas = executionGas + _getUnusedGasPenalty(executionGas, mUserOp.callGasLimit) + opInfo.preOpGas;
+            if (context.length > 0) {
```

Comment:
Also for the entire `if (context.length > 0) { ... }` block - it would probably cost nothing to make it an internal function.